mshadow/cuda/tensor_gpu-inl.cuh

```diff
@@ -488,6 +548,70 @@ inline void AddTakeGrad(Tensor<gpu, 2, DType> dst,
        src.size(0),
        src.size(1));
 }
+
+template<typename IndexType, typename DType>
+inline void AddTakeGradLargeBatch(Tensor<gpu, 2, DType> dst,
+                                  const Tensor<gpu, 1, IndexType>& sorted,
+                                  const Tensor<gpu, 1, IndexType>& index,
+                                  const Tensor<gpu, 2, DType> &src) {
+  const int kWarpBits = kMemUnitBits;
+  const int SZ = 4;
+  const int block_dim_x = 1 << kWarpBits;
+  const int block_dim_y = 4;
+  const int grid_dim_x = (src.size(0) + block_dim_y - 1) / block_dim_y;
+  const int grid_dim_y = (src.size(1) + block_dim_x * SZ - 1) / (block_dim_x * SZ);
+  dim3 dimBlock(block_dim_x, block_dim_y);
+  dim3 dimGrid(grid_dim_x, grid_dim_y);
+
+  CHECK_EQ(dst.size(1), src.size(1)) << "AddTakeGradLargeBatch: shape mismatch";
+  CHECK_EQ(index.size(0), src.size(0)) << "AddTakeGradLargeBatch: shape mismatch";
+  CheckLaunchParam(dimGrid, dimBlock, "AddTakeGradLargeBatch");
+  cudaStream_t stream = Stream<gpu>::GetStream(dst.stream_);
+
+  AddTakeGradLargeBatchKernel<kWarpBits, SZ, DType>
+      <<<dimGrid, dimBlock, 0, stream>>>
+      (dst.dptr_,
+       sorted.dptr_,
+       index.dptr_,
+       src.dptr_,
+       static_cast<int>(src.size(0)),
+       static_cast<int>(src.size(1)));
+}
+
+template<typename KDType, typename VDType>
+inline void SortByKey(Tensor<gpu, 1, KDType> keys, Tensor<gpu, 1, VDType> values,
+                      bool is_ascend) {
+  CHECK_EQ(keys.CheckContiguous(), true);
+  CHECK_EQ(values.CheckContiguous(), true);
+  cudaStream_t stream = Stream<gpu>::GetStream(keys.stream_);
+  thrust::device_ptr<KDType> key_iter = thrust::device_pointer_cast(keys.dptr_);
+  thrust::device_ptr<VDType> value_iter = thrust::device_pointer_cast(values.dptr_);
+  if (is_ascend) {
+    thrust::stable_sort_by_key(
+#if CUDA_VERSION >= 7000
+      thrust::cuda::par.on(stream),
```

Comment:
That's probably going to break dependency engine. You can either LOG(FATAL) and turn off the support or add cudastreamsync on the default stream